server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashCellIdSource.java

```diff
@@ -85,21 +85,42 @@ protected int advanceValue(org.elasticsearch.common.geo.GeoPoint target, int val
         @Override
         protected int advanceValue(org.elasticsearch.common.geo.GeoPoint target, int valuesIdx) {
             final String hash = Geohash.stringEncode(target.getLon(), target.getLat(), precision);
-            if (validHash(hash)) {
+            if (validPoint(target.getLon(), target.getLat()) || validHash(hash)) {
                 values[valuesIdx] = Geohash.longEncode(hash);
                 return valuesIdx + 1;
             }
             return valuesIdx;
         }
 
         private boolean validHash(String hash) {
-            final Rectangle rectangle = Geohash.toBoundingBox(hash);
+            final Rectangle rect = Geohash.toBoundingBox(hash);
+            if (rect.getMaxX() < rect.getMinX()) {
```

Comment:
This whole thing could really use a drawing in the comments 😄 .

Reply:
done